src/commands/compose/getComposeSubsetList.ts

```diff
@@ -0,0 +1,128 @@
+    const previousChoices = ext.context.workspaceState.get<string[]>(workspaceProfileListKey, profiles);
+    const result = await pickSubsets(context, 'profiles', profiles, previousChoices);
+
+    // Update the cache
+    await ext.context.workspaceState.update(workspaceProfileListKey, result);
+
+    return result.map(p => `--profile ${p}`).join(' ');
+}
+
+export async function getComposeServiceList(context: IActionContext, workspaceFolder: vscode.WorkspaceFolder, composeCommand: string): Promise<string> {
+    const services = await getServiceSubsets(workspaceFolder, composeCommand, 'services');
+
+    if (!services?.length) {
+        context.errorHandling.suppressReportIssue = true;
+        throw new Error(localize('vscode-docker.getComposeSubsetList.noServices', 'No services were found in the compose document(s). Did you mean to use profiles instead?'));
+    }
+
+    // Fetch the previously chosen services list. By default, all will be selected.
+    const workspaceServiceListKey = `vscode-docker.composeServices.${workspaceFolder.name}`;
+    const previousChoices = ext.context.workspaceState.get<string[]>(workspaceServiceListKey, services);
+    const result = await pickSubsets(context, 'services', services, previousChoices);
+
+    // Update the cache
+    await ext.context.workspaceState.update(workspaceServiceListKey, result);
+
+    return result.join(' ');
+}
+
+async function pickSubsets(context: IActionContext, type: SubsetType, allChoices: string[], previousChoices: string[]): Promise<string[]> {
+    const label = type === 'profiles' ?
+        localize('vscode-docker.getComposeSubsetList.chooseProfiles', 'Choose profiles to start') :
+        localize('vscode-docker.getComposeSubsetList.choose', 'Choose services to start');
+
+    const pickChoices: IAzureQuickPickItem<string>[] = allChoices.map(s => ({
+        label: s,
+        data: s,
+        picked: previousChoices.some(p => p === s),
+    }));
+
+    const chosenSubsets = await context.ui.showQuickPick(
+        pickChoices,
+        {
+            canPickMany: true,
+            placeHolder: label,
+        }
+    );
+
+    context.telemetry.measurements.totalServices = pickChoices.length;
+    context.telemetry.measurements.chosenServices = chosenSubsets.length;
+    context.telemetry.properties.subsetType = type;
+
+    return chosenSubsets.map(c => c.data);
+}
+
+async function getServiceSubsets(workspaceFolder: vscode.WorkspaceFolder, composeCommand: string, type: SubsetType): Promise<string[] | undefined> {
+    // TODO: if there are any profiles, then only services with no profiles show up when you query `config --services`. This makes for a lousy UX.
+
+    try {
+        // Start by getting a new command with the exact same files list (replaces the "up ..." or "down ..." with "config --services" or "config --profiles")
+        const configCommand = composeCommand.replace(composeCommandReplaceRegex, `config --${type}`);

```

Comment:
We should use JSON output.

Reply:
The only way to get all the information would be to use `docker-compose config --profiles` to get a list of profiles, then use `docker-compose config --format json` once for each profile, then dedupe. IMO that's just silly, so I'll leave it as it is now.